### PR TITLE
BUG: exp, log AVX loops do not use steps

### PR DESCRIPTION
**Pull request by r-devulap**

### CODE REVIEW

**Comment by r-devulap** [comment]
This passes the `numpy/core/tests/test_ufunc.py` and `numpy/core/tests/test_umath.py ` tests, but I am seeing non-deterministic failures in `numpy/core/tests/test_numeric.py`. It fails about 1 in 5 times I run the test suite (this happens even for the master branch). Yet to determine what is causing it.

**Comment by r-devulap** [comment]
Also, seems like reciprocal is failing the newly added test: `test_ufunc_noncontiguous`.

**Comment by mattip** [comment]
The reciprical failure is due to `1.0 / 0`, when cast to `npy_int`, resulting in either `-1` or `1 + 2**31`

Perhaps it is due to using `-O3` with the different branches in `UNARY_LOOP_FAST`. What happens if you make the numerator `1.0F` on line 717 in `loops.c.src`?

**Comment by r-devulap** [comment]
now the same test seems to fail for macOS int16 ...

**Comment by juliantaylor** [comment]
due to the new generic test?
if so one could remove the fast loop for that function and handle it in another issue
I don't think performance of integer reciprocal operations is particularly important

**Comment by mattip** [comment]
Agreed `np.reciprical(np.array([0], t) for t = [int dtypes]` should be a separate issue. Perhaps for now change `range(6)` to `range(1, 7)` to avoid 0

**Comment by r-devulap** [comment]
please do not merge this just yet. now might be a good time to think about any potential bugs/missing elements of this patch. @juliantaylor could you please review this when you have the time?

File: numpy/core/src/umath/simd.inc.src [hunk 1]
```diff
@@ -122,20 +122,36 @@ abs_ptrdiff(char *a, char *b)
...
  */
 
 /* prototypes */
-#if defined HAVE_ATTRIBUTE_TARGET_@ISA@_WITH_INTRINSICS && defined NPY_HAVE_SSE2_INTRINSICS
 
 /**begin repeat1
  * #func = exp, log#
  */
 
+#if defined HAVE_ATTRIBUTE_TARGET_@ISA@_WITH_INTRINSICS && defined NPY_HAVE_SSE2_INTRINSICS
```
Comment (juliantaylor) [inline numpy/core/src/umath/simd.inc.src]: why are SSE2 intrinsic needed in this avx2/avx512 section?

File: numpy/core/src/umath/loops.c.src [hunk 2]
```diff
@@ -1599,21 +1599,16 @@ FLOAT_@func@(char **args, npy_intp *dimensions, npy_intp *steps, void *NPY_UNUSE
 NPY_NO_EXPORT NPY_GCC_OPT_3 void
 FLOAT_@func@_@isa@(char **args, npy_intp *dimensions, npy_intp *steps, void *NPY_UNUSED(data))
 {
+    if (!run_unary_@isa@_@func@_FLOAT(args, dimensions, steps)) {
+        UNARY_LOOP {
 #if defined @CHK@ && defined NPY_HAVE_SSE2_INTRINSICS
-    @ISA@_@func@_FLOAT((npy_float*)args[1], (npy_float*)args[0], dimensions[0]);
+            @ISA@_@func@_FLOAT((npy_float *)op1, (npy_float *)ip1, 1);
```
Comment (juliantaylor) [inline numpy/core/src/umath/loops.c.src]: this is not correct, the existing logic is to dispatch to the vectorized code function if appropriate and if not run the fallback code inside the if statement.
a logic that may be unnecessarily convoluted, the simd code is in need of refactoring but that is a different issue. To fix your code just remove this line and the preprocessor directives around it.

the `NPY_GCC_OPT_3` macro can probably also be removed, the UNARY_LOOP macro does not allow for much optimization and to vectorize it with UNARY_LOOP_FAST gcc would require the `-ffast-math` which we intentionally do not use.

**Comment by juliantaylor** [comment]
so far I know it doesn't apply to exp and log but in general you also need to take care that scalar and non scalar code give the same results

for some functions we unfortunately use different code. So far I know this applies to for example `power`, see some old experiments of mine: https://github.com/juliantaylor/numpy/commit/6c29106dc05dfb64be711b1929dd9bfd280379bb

**Comment by juliantaylor** [comment]
With the commented issue and a clarification on one maybe unnecessary macro check this looks good

**Comment by r-devulap** [comment]
@mattip: Can we always expect the stride of input array in FLOAT_exp_AVX to be constant? what happens in this case?
```
>>> np.random.rand(10,5)[::2].strides
(80, 8)
```

**Comment by juliantaylor** [comment]
@r-devulap the strides are always constant for the ufunc inner loops.
In your strided case the iterator will copy to a contiguous buffer and then call the inner loop on the buffer

**Comment by juliantaylor** [comment]
Actually for these expensive functions it could be worthwhile to make the iterator always buffering, as it will be faster to copy to a buffer and call the vectorized function than using the unvectorized one
it also ensures the same results for all strides of data

but I am not sure if we can configure that on a per ufunc level

**Comment by juliantaylor** [comment]
It is very worthwhile:

```
import numpy as np
d = np.random.rand(1000, 50).astype(np.float32)
print("2d strided -> buffered iterator")
%timeit np.exp(d[::2])
d = np.random.rand(1000* 50).astype(np.float32)
print("1d strided unbuffered iterator")
%timeit np.exp(d[::2])


2d strided -> buffered iterator
10000 loops, best of 3: 77.3 µs per loop
1d strided unbuffered iterator
1000 loops, best of 3: 1.02 ms per loop
```

So this is something we should look into before release, in particular the same result for different strided data is important. We do have differences with the summation depending on strides and that issues comes up all the time.

**Comment by juliantaylor** [comment]
opened https://github.com/numpy/numpy/issues/13557 for buffered iterator and result consistency restoring.

**Comment by seberg** [comment]
I do not think you can do this always, but I have seen that you (Julian, at least I think so) added an iterator flags (and op flags) to the ufunc struct. So, for most loops using `ufunc->op_flags = {NPY_ITER_CONTIG, NPY_ITER_CONTIG, NPY_ITER_CONTIG};`  (or similar).

But I am not sure they are picked up for all things such as `reduceat` or `accumulate`.

**Comment by juliantaylor** [comment]
wait I forgot this branch still has the bug I commented on when I benchmarked ...
Actually I see no performance difference when using correct. Turns out glibc expf is actually faster than our avx2 vectorized version when called directly as the specialized ufunc now does.

So we should probably take a step back and do some platform benchmarking to see how many platforms actually already provide a well performing function for us.

But this fix (with the commented issue) can already be merged so we have correct results and benchmarks again.

**Comment by mattip** [comment]
@juliantaylor this PR still needs the fix from [your comment](https://github.com/numpy/numpy/pull/13520#discussion_r283136057), correct?

**Comment by juliantaylor** [comment]
yes

**Comment by mattip** [comment]
We are getting close to the deadline for 1.17.0. We should finish this so it gets some testing before RC1

**Comment by juliantaylor** [comment]
hm so soon already? This is a very significant low level change and it should not be done hastily.
Also with this fix I would not consider it finished. At least we need to ensure equal results regardless of the strides of the input and I would still like to see some benchmarks on more platforms comparing to system libraries.

**Comment by r-devulap** [comment]
@juliantaylor @mattip I am slightly unclear about this comment https://github.com/numpy/numpy/pull/13520#discussion_r283136057. Removing those 3 lines will result in your new test failing because strided and non-strided outputs will be different. I think we need to keep these lines. Am I missing something?

**Comment by r-devulap** [comment]
> Actually for these expensive functions it could be worthwhile to make the iterator always buffering, as it will be faster to copy to a buffer and call the vectorized function than using the unvectorized one
> it also ensures the same results for all strides of data
> 
> but I am not sure if we can configure that on a per ufunc level

Instead of copying entire array to a buffer and call the vectorized function, won't it be better to use _mm256_mask_i32gather_ps instructions to accommodate strides in the vectorized function? Since stride is always uniform this is fairly easy to do and might speed up performance of the strided case.

**Comment by juliantaylor** [comment]
oh I misinterpreted the intention of the line, yes it is needed for that purpose. Please add a comment to the line why it exists.
That it causes a factor 10 slowdown for strided data is still a concern.

Using gather would be an option to use the same code with lower impact, but the performance vs a buffered iterator should be tested.
Implementing the gather in the iterator buffer creating would be better as that would benefit all ufuncs not only float32 exp/log but as a quick fix one could also do it in those functions and iterate on the solution later.

**Comment by juliantaylor** [comment]
The single element loop is still an issue as the code will currently loads at least 32 bytes of data on a single element loop, but there is no guarantee that the input data is 32 byte aligned so it can cross a page boundary and cause a segmentation fault.
This could be fixed with buffering too.

**Comment by r-devulap** [comment]
sounds good to me. I am working on using gather instructions for exp and log, but will need some time to implement, test and benchmark. May be we can merge this as is and I can submit another patch with gather instructions? would that be okay?

**Comment by juliantaylor** [comment]
or by adapting the load_partial_lanes function to handle the unaligned case via scalar loads

**Comment by r-devulap** [comment]
loading 32 bytes of data even across a page boundary is not a problem with the masked and unaligned load and stores. that's what the masks are for, right?

**Comment by juliantaylor** [comment]
A I did not know that, looking at the description it does say the masked value is not loaded, though it does not explicitly say the masked memory may be unmapped and will not throw and exception but that could be inferred from the wording. If that does indeed work then the code is fine.

**Comment by juliantaylor** [comment]
I still have the one comment on the sse2 preprocessor but that does no harm so ok from me for merging

but we should probably still discuss release timing in the community meeting tomorrow (if there is one).

**Comment by r-devulap** [comment]
I added in a comment stating the performance issues for strided loops. I will submit another patch to fix that using gather instructions (if it fixes the performance issues). I think this is okay to merge now.

**Comment by mattip** [comment]
There are merge conflicts.

**Comment by r-devulap** [comment]
re-based with master, no conflicts now.

**Pull request merged**